### PR TITLE
[Feat]: year calendar apis

### PR DESCRIPTION
**Pull request by kalash-copods**
created update api and modified create and get api for year calendar.
modified year calendar schema and serializers

### CODE REVIEW

File: LeaveTrackingApp/views/calendar_and_policy_views.py [hunk 1]
```diff
@@ -6,36 +6,119 @@
...
+            #publish calendar
+            elif calendar_obj.status == 'approved' and document_data['status'] == 'Published':
+                calendar_obj.status = 'published'
+                calendar_obj.save()
+                return JsonResponse({'message': 'calendar published'}, status=status.HTTP_200_OK)
+
+        except Exception as e:
+            return JsonResponse({"error": str(e)}, status=status.HTTP_500_INTERNAL_SERVER_ERROR)
     
+
```
Comment (richa-copods) [inline LeaveTrackingApp/views/calendar_and_policy_views.py:121]: Remove extra line

**Review by richa-copods — CHANGES_REQUESTED** [review]

File: LeaveTrackingApp/serializers/calendar_and_policy_serializers.py [hunk 2]
```diff
@@ -43,6 +39,33 @@ def create(self, validated_data):
 
         return holiday_calendar
 
+    @transaction.atomic
+    def update(self, instance, validated_data):
+        validated_data['status'] = 'draft' # set status to draft when policy gets edited
+        holiday_data = validated_data.pop('holidays', None)
+        print(holiday_data)
```
[the remark this answers is not in the document]
Comment (kalash-copods) [reply LeaveTrackingApp/serializers/calendar_and_policy_serializers.py]: resolved

File: LeaveTrackingApp/views/calendar_and_policy_views.py [hunk 3]
```diff
@@ -6,36 +6,119 @@
...
-        year = request.GET.get('year', None)
-        if year:
-            holiday_calendars = yearCalendar.objects.filter(status=status, year=year)
-        else:
-            holiday_calendars = yearCalendar.objects.filter(status=status)
-        holiday_calendars_serializer = YearCalendarSerializerList(holiday_calendars, many=True)
-        return JsonResponse(holiday_calendars_serializer.data, safe=False)
+        try:
+            year = request.GET.get('year', None)
+            if year:
```
[the remark this answers is not in the document]
Comment (kalash-copods) [reply LeaveTrackingApp/views/calendar_and_policy_views.py:20]: already handled it in create api

File: LeaveTrackingApp/views/calendar_and_policy_views.py [hunk 4]
```diff
@@ -6,36 +6,119 @@
...
+                year_calendar = yearCalendar.objects.filter(status='published', year=year).order_by('-updated_at').first()
+            else:
+                year_calendar = yearCalendar.objects.filter(status='published').order_by('-updated_at').first()
+            if not year_calendar:
+                if year:
+                    year_calendar = yearCalendar.objects.filter(status__in=['approved', 'draft', 'sent_for_approval'], year=year).order_by('-updated_at').first()
```
[the remark this answers is not in the document]
Comment (kalash-copods) [reply LeaveTrackingApp/views/calendar_and_policy_views.py]: resolved

File: LeaveTrackingApp/views/calendar_and_policy_views.py [hunk 5]
```diff
@@ -6,36 +6,119 @@
...
 @user_is_authorized
 def createHolidayCalendar(request):
     if request.method=='POST':
-        holiday_calendar_data = JSONParser().parse(request)
-        holiday_calendar_serializer = YearCalendarSerializer(data=holiday_calendar_data)
-        if holiday_calendar_serializer.is_valid():
-            holiday_calendar_serializer.save()
-            return JsonResponse(holiday_calendar_serializer.data, status=status.HTTP_201_CREATED)
-        return JsonResponse(holiday_calendar_serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+        try:
```
[the remark this answers is not in the document]
Comment (kalash-copods) [reply LeaveTrackingApp/views/calendar_and_policy_views.py:40]: will be handled in frontend by clearing the populated data and edit the existing one only (suggested by Aditya), as we are having only one draft state, deleting from db would be of no use

File: LeaveTrackingApp/views/calendar_and_policy_views.py [hunk 6]
```diff
@@ -6,36 +6,119 @@
...
+            year_calendar_data = JSONParser().parse(request)
+            if yearCalendar.objects.filter(status__in=['approved', 'draft', 'sent_for_approval']).exists():
+                return JsonResponse({"error": "A draft or an approved calendar already exists"}, status=status.HTTP_403_FORBIDDEN)
+            
+            if not year_calendar_data.get('status') or year_calendar_data['status'] != 'Draft':
```
[the remark this answers is not in the document]
Comment (kalash-copods) [reply LeaveTrackingApp/views/calendar_and_policy_views.py]: resolved

File: LeaveTrackingApp/views/calendar_and_policy_views.py (hunk 1 above)
Comment (kalash-copods) [reply LeaveTrackingApp/views/calendar_and_policy_views.py]: resolved

**Review by richa-copods — APPROVED** [review]

**Pull request merged**